File: 0155_min_stack/min_stack.go
package main

type Stack struct {
	val int
	min int
}

type MinStack struct {
	s []Stack
}

func Constructor() MinStack {
	return MinStack{}
}

func (this *MinStack) Push(x int) {
	if len(this.s) == 0 {
		this.s = append(this.s, Stack{x, x})
	} else {
		minVal := this.GetMin()
		this.s = append(this.s, Stack{x, Min(x, minVal)})
	}
}

func (this *MinStack) Pop() {
	this.s = this.s[:len(this.s)-1]
}

func (this *MinStack) Top() int {
	return this.s[len(this.s)-1].val
}

func (this *MinStack) GetMin() int {
	if len(this.s) > 0 {
		return this.s[len(this.s)-1].min
	}
	return 0
}

func Min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */

func main() {

}

File: 0155_min_stack/min_stack.java
import java.util.Stack;
import javafx.util.Pair;

class MinStack {
    private Stack<Pair<Integer, Integer>> myStack;

    /** initialize your data structure here. */
    public MinStack() {
        myStack = new Stack();
    }

    public void push(int x) {
        int tmp = myStack.isEmpty() ? x : myStack.peek().getValue();
        myStack.push(new Pair(x, Math.min(x, tmp)));
    }

    public void pop() {
        myStack.pop();
    }

    public int top() {
        return myStack.peek().getKey();
    }

    public int getMin() {
        return myStack.peek().getValue();
    }
}

/**
 * Your MinStack object will be instantiated and called as such: MinStack obj =
 * new MinStack(); obj.push(x); obj.pop(); int param_3 = obj.top(); int param_4
 * = obj.getMin();
 */